cluster-autoscaler/core/scale_down.go

```diff
@@ -103,6 +103,93 @@ func (n *NodeDeleteStatus) SetDeleteInProgress(status bool) {
 	n.deleteInProgress = status
 }
 
+type resourceName string
+type scaleDownResourcesLimits map[resourceName]int64
+type scaleDownResourcesDelta map[resourceName]int64
+
+func computeScaleDownResourcesLeftLimits(nodes []*apiv1.Node, cp cloudprovider.CloudProvider, timestamp time.Time) (scaleDownResourcesLimits, errors.AutoscalerError) {
+	resourceLimiter, err := cp.GetResourceLimiter()
+	if err != nil {
+		return scaleDownResourcesLimits{}, errors.ToAutoscalerError(
+			errors.CloudProviderError,
+			err)
+	}
+
+	totals := make(map[resourceName]int64)
+	totals[cloudprovider.ResourceNameCores], totals[cloudprovider.ResourceNameMemory] = calculateCoresAndMemoryTotal(nodes, timestamp)
+
+	for gpuName, gpuCount := range calculateGpusTotal(nodes, cp, timestamp) {
+		totals[resourceName(gpuName)] = gpuCount
+	}
+
+	resultScaleDownLimits := make(scaleDownResourcesLimits)
+	for _, resource := range resourceLimiter.GetResources() {
+		total := totals[resourceName(resource)]
+		min := resourceLimiter.GetMin(resource)
+
+		// we put only actual limits into final map. No entry means no limit.
+		if min > 0 {
+			resultScaleDownLimits[resourceName(resource)] = total - min
+		}
+	}
+	return resultScaleDownLimits, nil
+}
+
+func noLimitsOnResources() scaleDownResourcesLimits {
+	return nil
+}
+
+func copyScaleDownResourcesLimits(source scaleDownResourcesLimits) scaleDownResourcesLimits {
+	copy := scaleDownResourcesLimits{}
+	for k, v := range source {
+		copy[k] = v
+	}
+	return copy
+}
+func computeScaleDownResourcesDelta(node *apiv1.Node, nodeGroup cloudprovider.NodeGroup) (scaleDownResourcesDelta, errors.AutoscalerError) {
+	nodeCPU, nodeMemory, err := getNodeCoresAndMemory(node)
+	if err != nil {
+		return scaleDownResourcesDelta{}, errors.ToAutoscalerError(errors.ApiCallError, err)
+	}
+
+	gpuType, gpuCount, err := gpu.GetNodeTargetGpus(node, nodeGroup)
+	if err != nil {
+		return scaleDownResourcesDelta{}, errors.ToAutoscalerError(errors.ApiCallError, err).AddPrefix("Failed to get node %v gpu: %v", node.Name)
+	}
+
+	return scaleDownResourcesDelta{
+		cloudprovider.ResourceNameCores:  nodeCPU,
+		cloudprovider.ResourceNameMemory: nodeMemory,
+		resourceName(gpuType):            gpuCount,
+	}, nil
+}
+
+func (limits *scaleDownResourcesLimits) checkDeltaWithinLimits(delta scaleDownResourcesDelta) *resourceName {
+	for resource, resourceDelta := range delta {
+		resourceLeft, found := (*limits)[resource]
+		if found {
+			if resourceDelta > resourceLeft {
+				return &resource
```

Comment:
In scale-up we want to treat no GPU limit as 0 and never add a node with that GPU. It is less obvious for scale-down though. Can we delete a node with GPU if no limit at all was specified? I slightly lean towards "yes", but it should be a very conscious decision. WDYT @mwielgus @aleksandra-malinowska?

Reply:
I don't think we always want to enforce specifying resource limits for GPUs in scale-up. Right now they're optional and it may surprise the users badly if autoscaling stops working for node groups with GPUs until they're set.

Reply:
> In scale-up we want to treat no GPU limit as 0 and never add a node with that GPU.

This is weird. 
I agree with what @aleksandra-malinowska wrote.